staging/vhost-device-can/src/vhu_can.rs

```diff
@@ -0,0 +1,2181 @@
+        let mut req_rx_buf = false;
+
+        for desc_chain in requests {
+            let atomic_mem = self.mem.as_ref().unwrap().memory();
+            let mut writer = desc_chain
+                .clone()
+                .writer(&atomic_mem)
+                .map_err(|_| Error::DescriptorWriteFailed)?;
+
+            let response = match self.controller.write().unwrap().pop() {
+                Ok(item) => {
+                    // If one or more frames have been received then request
+                    // a new rx buffer.
+                    req_rx_buf = true;
+                    item
+                }
+                Err(QueueEmpty) => {
+                    trace!("Empty queue!");
+                    return Ok(req_rx_buf);
+                }
+                Err(_) => {
+                    trace!("Pop error!");
+                    return Err(Error::HandleEventUnknown);
+                }
+            };
+
+            let can_rx = match self.check_rx_frame(response) {
+                Ok(frame) => frame,
+                Err(e) => {
+                    warn!("The tx frame had the following error: {}", e);
+                    return Err(e);
+                }
+            };
+
+            writer
+                .write_obj(can_rx)
+                .map_err(|_| Error::DescriptorWriteFailed)?;
+
+            if vring
+                .add_used(desc_chain.head_index(), writer.bytes_written() as u32)
+                .is_err()
+            {
+                warn!("Couldn't return used descriptors to the ring");
+            }
+        }
+
+        Ok(true)
+    }
+
+    /// Process the messages in the vring and dispatch replies
+    fn process_ctrl_queue(&mut self, vring: &VringRwLock) -> Result<()> {
+        trace!("process_ctrl_queue");
+        let requests: Vec<CanDescriptorChain> = vring
+            .get_mut()
+            .get_queue_mut()
+            .iter(self.mem.as_ref().unwrap().memory())
+            .map_err(|_| Error::DescriptorNotFound)?
+            .collect();
+
+        if self.process_ctrl_requests(requests, vring)? {
+            // Send notification once all the requests are processed
+            vring
+                .signal_used_queue()
+                .map_err(|_| Error::NotificationFailed)?;
+        }
+        Ok(())
+    }
+
+    /// Process the messages in the vring and dispatch replies
+    fn process_tx_queue(&self, vring: &VringRwLock) -> Result<()> {
+        trace!("process_tx_queue");
+        let requests: Vec<CanDescriptorChain> = vring
+            .get_mut()
+            .get_queue_mut()
+            .iter(self.mem.as_ref().unwrap().memory())
+            .map_err(|_| Error::DescriptorNotFound)?
+            .collect();
+
+        if self.process_tx_requests(requests, vring)? {
+            // Send notification once all the requests are processed
+            vring.signal_used_queue().map_err(|_| {
+                trace!("signal_used_queue error");
```

Comment:
I think you can remove this `trace!` as it is in other parts of code.

Reply:
No to be honest, some of those traces left there only for debugging purposes. I have removed them in the latest version, and kept only the ones which are needed for the user.